presto-docs/src/main/sphinx/security/built-in-system-access-control.rst

```diff
@@ -133,8 +140,54 @@ and Kerberos authentication:
           "allow": true
         }
       ],
-      "user_patterns": [
-        "(.*)",
-        "([a-zA-Z]+)/?.*@.*"
+      "principals": [
+        {
+          "principal": "(.*)",
+          "allow": true
+        },
+        {
+          "principal": "([a-zA-Z]+)/?.*@.*",
+          "allow": true
+        }
+      ]
+    }
+
+The following implements a matching that the principal name is the prefix of the
+username:
+
+.. code-block:: json
+
+    {
+      "catalogs": [
+        {
+          "allow": true
+        }
+      ],
+      "principals": [
+        {
+          "principal": "(.*)@example.net",
+          "user": "$1.*",
```

Comment:
That means the algorithm is:
- take principal
- match `principal` pattern, if it matches
- produce some string using `user` as substitution (substitution pattern)
- take this string and _treat it as another regex, match it against_ requested user name

This last step is not obvious. Would it be enough if we replaced the last step with:
- take this string and check if it is _equal to_ requested user name
?

...
The example of `privileged_service@example.net` principal below suggest that simplification wouldn't be OK.
However, what's going to happen when principal contains some non-ascii characters? Like `|.*|`? (Probably not possible, but still, this looks like a security issue)

Reply:
Maybe the best approach would be to separate this two use-cases. Like (using yaml notation)
```
# This addresses the use-case: principal name is not necessary to be related to the username and we sometimes want to share the same principle across multiple users
# here, users is just a static list of allowed users, which is that *contains*-checked against requested user name
# (principal can still be a pattern)
- principal: "priviledged@example.net"
  users: [bob, alice, findepi]

# This addresses the use-case: each user has its own principal and this is how they are expected to connect to presto
# here, user_pattern is meant to be "substitution pattern" which is than *equals*-checked against requested user name (or maybe name this `user_extraction` or something?)
- principal: "(\w*)@example.net"
  user_pattern: "$1"
```

This avoids applying `Pattern.compile` to something dynamically constructed based on external input.

Would that satisfy all requirements?

Reply:
I understand the biggest concern is using an external string as part of the regex. I think we can simply quote the group values in the regex?

Reply:
I just realized the group value could also include `\E` which is the sign of the end of quote.

Reply:
Well, technically, you could do manual substitution: replace each `$n` with `Pattern.quote(matcher.group(n))`.  (just need to take care not to subst `\$n`, this again is not hard to do)

I would like, however, to avoid all this, if possible.